botocore/credentials.py

```diff
@@ -436,11 +444,169 @@ def get_frozen_credentials(self):
+            Possible keys include, but may not be limited to,
+            DurationSeconds, Policy, SerialNumber, ExternalId and
+            RoleSessionName.
+
+        :type mfa_prompter: callable
+        :param mfa_prompter: A callable that returns input provided by the
+            user (i.e raw_input, getpass.getpass, etc.).
+
+        :type cache: dict
+        :param cache: An object that supports ``__getitem__``,
+            ``__setitem__``, and ``__contains__``.  An example of this is
+            the ``JSONFileCache`` class in aws-cli.
+
+        :type expiry_window_seconds: int
+        :param expiry_window_seconds: The amount of time, in seconds,
+        """
+        self._client_creator = client_creator
+        self._source_credentials = source_credentials
+        self._role_arn = role_arn
+        self._expiry_window_seconds = expiry_window_seconds
+
+        if extra_args is None:
+            self._assume_kwargs = {}
+        else:
+            self._assume_kwargs = deepcopy(extra_args)
+        self._assume_kwargs['RoleArn'] = self._role_arn
+
+        self._role_session_name = self._assume_kwargs.get('RoleSessionName')
+        self._using_default_session_name = False
+        if not self._role_session_name:
+            self._role_session_name = 'botocore-session-%s' % (
+                int(time.time()))
+            self._assume_kwargs['RoleSessionName'] = self._role_session_name
+            self._using_default_session_name = True
+
+        self._mfa_prompter = mfa_prompter
+        if self._mfa_prompter is None:
+            self._mfa_prompter = getpass.getpass
+
+        self._cache = cache
+        if self._cache is None:
+            self._cache = {}
+
+        self._cache_key = self._create_cache_key()
+
+    def _create_cache_key(self):
+        """Create a predictable cache key for the current configuration.
+
+        The cache key is intended to be compatible with file names.
+        """
+        # On windows, ':' is not allowed in file names, so we'll
+        # replace them with '_' instead.
+        cache_key = self._role_arn.replace(':', '_').replace(os.path.sep, '_')
+
+        if not self._using_default_session_name:
+            cache_key += '--%s' % self._role_session_name
+
+        policy = self._assume_kwargs.get('Policy')
+        if policy:
+            # To have a predictable hash, the keys of the policy must be
+            # sorted.
+            policy = json.dumps(json.loads(policy), sort_keys=True)
+            cache_key += '--%s' % sha256(policy.encode()).hexdigest()
+
+        return cache_key
+
+    def fetch_credentials(self):
+        return self._get_cached_credentials()
+
+    def _get_cached_credentials(self):
+        """Get up-to-date credentials.
+
+        This will check the cache for up-to-date credentials, calling assume
+        role if none are available.
+        """
+        response = self._load_from_cache()
```

Comment:
The original version had debug logs about whether or not we load from cache that don't appear to be ported over with this change (line 856 in the original file).

It would be helpful to log when we load from the cache and when we load from cache but they're expired.

Reply:
👍